File: packages/shared-types/src/category.ts
import type { Timestamp } from './index'

/**
 * Document /categories/{categoryId} — référentiel d'âge éditable par l'admin.
 * Voir docs/firebase.md (/categories) et docs/main.md (section "Catégories d'équipes").
 *
 * Référencé par /teams.categoryId. Le nom et la tranche d'âge ne sont PAS
 * dénormalisés sur /teams : un rename se reflète automatiquement.
 */
export interface CategoryData {
  /** Libellé affiché : "U14", "Seniors A", "Loisirs". */
  name: string
  /** Borne basse (incluse). null = catégorie ouverte par le bas. */
  minAge: number | null
  /** Borne haute (incluse). null = catégorie ouverte par le haut. */
  maxAge: number | null
  /** Entier croissant pour tri stable dans pickers. Tie-break : minAge puis name. */
  displayOrder: number
  /** false = archivée (n'apparaît plus dans le picker, teams existantes inchangées). */
  active: boolean
  createdAt: Timestamp
}

export type Category = CategoryData & { id: string }

File: apps/mobile/lib/models/category.dart
import 'package:cloud_firestore/cloud_firestore.dart';

import '../core/firestore_converters.dart';

/// Document `/categories/{categoryId}` — référentiel d'âge éditable par l'admin.
class Category {
  const Category({
    required this.id,
    required this.name,
    required this.minAge,
    required this.maxAge,
    required this.displayOrder,
    required this.active,
    required this.createdAt,
  });

  final String id;
  final String name;

  /// Borne basse incluse. `null` = catégorie ouverte par le bas.
  final int? minAge;

  /// Borne haute incluse. `null` = catégorie ouverte par le haut.
  final int? maxAge;
  final int displayOrder;
  final bool active;
  final DateTime createdAt;

  /// Libellé de la tranche d'âge dérivé des bornes.
  String get ageRangeLabel {
    if (minAge == null && maxAge == null) return 'Ouvert';
    if (minAge != null && maxAge == null) return '$minAge ans+';
    if (minAge != null && maxAge != null) {
      return minAge == maxAge ? '$minAge ans' : '$minAge-$maxAge ans';
    }
    return 'Jusqu\'à $maxAge ans';
  }

  factory Category.fromMap(String id, Map<String, dynamic> map) => Category(
        id: id,
        name: FsConvert.str(map['name']),
        minAge: FsConvert.intOrNull(map['minAge']),
        maxAge: FsConvert.intOrNull(map['maxAge']),
        displayOrder: FsConvert.integer(map['displayOrder']),
        active: FsConvert.boolean(map['active'], true),
        createdAt: FsConvert.toDateTimeOr(map['createdAt']),
      );

  factory Category.fromFirestore(
    DocumentSnapshot<Map<String, dynamic>> snapshot,
  ) =>
      Category.fromMap(snapshot.id, FsConvert.requireData(snapshot));

  Category copyWith({
    String? name,
    int? minAge,
    int? maxAge,
    int? displayOrder,
    bool? active,
    DateTime? createdAt,
  }) =>
      Category(
        id: id,
        name: name ?? this.name,
        minAge: minAge ?? this.minAge,
        maxAge: maxAge ?? this.maxAge,
        displayOrder: displayOrder ?? this.displayOrder,
        active: active ?? this.active,
        createdAt: createdAt ?? this.createdAt,
      );
}
